lib/EarlyFunctionAnalysis/UsedRegisters.h

```diff
@@ -0,0 +1,152 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <map>
+#include <set>
+
+#include "revng/EarlyFunctionAnalysis/CallGraph.h"
+#include "revng/EarlyFunctionAnalysis/Common.h"
+#include "revng/MFP/SetLattices.h"
+#include "revng/Model/Register.h"
+#include "revng/Support/MetaAddress.h"
+
+#include "DetectABI.h"
+
+namespace efa {
+
+struct FunctionABI {
+  using SetOfRegisters = std::set<model::Register::Values>;
+  using SUL = SetUnionLattice<SetOfRegisters>;
+  SetOfRegisters ArgumentRegisters;
+  SetOfRegisters ReturnRegisters;
+};
+
+// using LatticeElement = std::map<MetaAddress, FunctionABI>;
+
+struct LatticeElement : std::map<MetaAddress, FunctionABI> {
+  virtual ~LatticeElement() = default;
+
+  [[nodiscard]] virtual bool
+  hasRetReg(MetaAddress Addr, model::Register::Values V) const {
+    if (count(Addr)) {
+      return at(Addr).ReturnRegisters.count(V);
+    }
+
+    return false;
+  }
+
+  [[nodiscard]] virtual bool
+  hasArgReg(MetaAddress Addr, model::Register::Values V) const {
+    if (count(Addr)) {
+      return at(Addr).ArgumentRegisters.count(V);
+    }
+
+    return false;
+  }
+};
+
+struct ExtremalLatticeElement : LatticeElement {
+  [[nodiscard]] bool
+  hasRetReg(MetaAddress Addr, model::Register::Values V) const {
+    return true;
+  }
+
+  [[nodiscard]] bool
+  hasArgReg(MetaAddress Addr, model::Register::Values V) const {
+    return true;
+  }
+};
+
+struct UsedRegistersMFI {
+  using Label = BasicBlockNode *;
+  using GraphType = llvm::Inverse<BasicBlockNode *>;
+  using LatticeElement = LatticeElement;
+  using ExtremalLatticeElement = ExtremalLatticeElement;
+
+  DetectABI &DA;
+
+  explicit UsedRegistersMFI(DetectABI &DA) : DA{ DA } {}
+
+  [[nodiscard]] LatticeElement
+  combineValues(const LatticeElement &E1, const LatticeElement &E2) const {
```

Comment:
Implement `FunctionABI::combineValues` and `LatticeElement::combineValues`: `UsedRegistersMFI::combineValues` should use `LatticeElement::combineValues`, which in turn should use `FunctionABI::combineValues`.

Same of `isLessThenOrEqual`.